[3DS] Enable CPU filters when loading Game Gear content

jdgleaver opened the pull request:
Currently, the core produces a black screen when attempting to use CPU filters on the 3DS (issue #156).

It turns out that this is a 3DS hardware limitation - it just doesn't have enough heap space. The 'problem' is that the core sets a very large max width/height in retro_get_system_av_info(). All the upscaling CPU filters double these values when allocating a frame buffer, so we end up with 1440x1152 - this is huge! The 3DS can't handle it, so no texture gets uploaded and we get a blank screen (at least it doesn't crash, which is nice...).

The only way around this is to reduce the max width/height. This is a little sticky, since the core emulates multiple systems and things become convoluted with in-game resolution changes and various aspect ratio shenanigans. Since Genesis and Master System games don't really benefit from CPU filters anyway (native resolution is a little too high, and 1:1 pixel mapping works nicely on the 240p 3DS screen), I though it best to avoid extensive changes to the code. Instead I focussed on the Game Gear, which is the only system that actively *needs* filtering, and which has a guaranteed 'safe' resolution (i.e. it can't ever exceed the heap space).

This pull request therefore does one very tiny thing: if (platform == 3DS) and (emulated system == Game Gear), the system_av_info max width/height are set to the actual emulated system screen dimensions. I understand that this smells like a hack, but as far as I can tell it is completely safe, and CPU filters can now be used.

So instead of this, which is disgusting:

![royal stone japan en v1 00 gg -181105-144409](https://user-images.githubusercontent.com/38211560/48008808-48181980-e112-11e8-8d6c-185572694879.png)

We can have something like this (Normal2x):

![royal stone japan en v1 00 gg -181105-144422](https://user-images.githubusercontent.com/38211560/48009086-db514f00-e112-11e8-98da-caa4407eeafa.png)

...or this (Scale2x):

![royal stone japan en v1 00 gg -181105-144435](https://user-images.githubusercontent.com/38211560/48009092-dd1b1280-e112-11e8-923b-0375edc8b252.png)

...both of which are markedly superior to the output of the official 3DS Game Gear Virtual Console emulator. (And the core is so efficient that all games run at full speed with Scale2x even on an o3DS!)

As an aside, it would probably be a good idea to modify the CTR video driver so CPU filters are automatically disabled if their use would exceed the heap space. I will try to look at this if I have time...

ekeeke commented:
I see potential issues with this hack unfortunately, which should have been taken intoconsideration before merging that PR :  if you change any core settings that increase fb_width or fb_height (namely "borders", "region", "ntsc filter" or "GG extended screen" settings) after loading a game gear game, it will likely result in a memory crash (or some black screen depending how safe is video drivers and CPU filters code in retroarch) because allocated video buffer size is smaller than rendered screen size. 

Also, libretro core are supposed to be platform-independent so I think it's a bad idea to add platform-specific hacks that break libretro API specification in libretro core  (max_width/max_height are supposed to be maximal screen size usable by loaded game, not some arbitrary default size that is used on reset)  or add some hardcoded core limitations because of some specific hardware use case (people could want use aforementioned core settings and not retroarch CPU filters but this hack limits their use of the core). It would have been better to deal with this on retroarch side (for example by making CPU filter or video drivers only allocate necessary memory when needed by directly using fb_width/fb_height instead of max values, for platforms that have such memory restrictions). This would surely have benefited other cores with similar behavior or other platforms with similar limitations.

Another alternative would be to drop these 3DS and SYSTEM_GG specific conditions and simply distinguish between 8-bit and 16-bit system emulation when returning max_width/max_height since this is specific to the loaded game and can not be changed after game is loaded. Interlaced video height being limited to 16-bit games, max_width and max_height could actually be limited to 720x288 for 8-bit games. a few more memory could even be saved by using exact needed width instead of some rounded approximation (maximal case being PAL region with NTSC filter and all borders enabled) i.e 668x288 for 8-bit games (that are limited to 256 pixels width mode) and 696x576 for 16-bit games. To distinguish between 8-bit and16-bit games, just check following condition (16-bit if True):   if ((system_hw & SYSTEM_PBC) == SYSTEM_MD)

jdgleaver commented:
@ekeeke: Thank you for your work on this fantastic emulator, and I apologise for any upset that has been caused!

> I see potential issues with this hack unfortunately, which should have been taken intoconsideration before merging that PR : if you change any core settings that increase fb_width or fb_height (namely "borders", "region", "ntsc filter" or "GG extended screen" settings) after loading a game gear game, it will likely result in a memory crash (or some black screen depending how safe is video drivers and CPU filters code in retroarch) because allocated video buffer size is smaller than rendered screen size.

No, this doesn't happen. RETRO_ENVIRONMENT_SET_SYSTEM_AV_INFO gets called, so any buffers are recreated with the correct size.

> Also, libretro core are supposed to be platform-independent so I think it's a bad idea to add platform-specific hacks that break libretro API specification in libretro core (max_width/max_height are supposed to be maximal screen size usable by loaded game, not some arbitrary default size that is used on reset) or add some hardcoded core limitations because of some specific hardware use case (people could want use aforementioned core settings and not retroarch CPU filters but this hack limits their use of the core). It would have been better to deal with this on retroarch side (for example by making CPU filter or video drivers only allocate necessary memory when needed by directly using fb_width/fb_height instead of max values, for platforms that have such memory restrictions). This would surely have benefited other cores with similar behavior or other platforms with similar limitations.

I agree 100%. The best solution would be an overhaul of the CPU filter interface (as you say), such that actual rather than maximum frame buffer size is used. The issue is that this is a lot of work, and I doubt anyone wants to do it... (the affected platforms are too niche, I think)

I do not like this little hack that I have implemented, but it does not limit use of the core, and without it Game Gear games are essentially unplayable on the 3DS (until you have seen it on real hardware, it is difficult to appreciate how disgusting unfiltered Game Gear games look on a 3DS XL screen... This is why M2 went to such extreme lengths with their filtering when implementing the official 3DS Game Gear Virtual Console emulator...)

> Another alternative would be to drop these 3DS and SYSTEM_GG specific conditions and simply distinguish between 8-bit and 16-bit system emulation when returning max_width/max_height since this is specific to the loaded game and can not be changed after game is loaded. Interlaced video height being limited to 16-bit games, max_width and max_height could actually be limited to 720x288 for 8-bit games. a few more memory could even be saved by using exact needed width instead of some rounded approximation (maximal case being PAL region with NTSC filter and all borders enabled) i.e 668x288 for 8-bit games (that are limited to 256 pixels width mode) and 696x576 for 16-bit games. To distinguish between 8-bit and16-bit games, just check following condition (16-bit if True): if ((system_hw & SYSTEM_PBC) == SYSTEM_MD)

This is a very good suggestion, and I like it. I will check whether this falls within the 3DS heap limit, and if so I'll make the relevant pull request.

(I've already modified the libretro video driver so CPU filters are disabled if there isn't enough heap space on the 3DS, but I want to test this a little more. If we can still use filters with a 668x288 frame buffer then we are golden, and this can all be handled properly without 'hacks')

jdgleaver commented:
> In that case, it should be possible to even more reduce max_width/max_height depending on the state of config.ntsc, config.overscan, config.render and config.gg_extra settings.

Wow - this is fantastic!

The 8 bit maximum of 668x288 is still too large for the 3DS heap, but with your fine-grained max_width/max_height specification set according to system/user config we can now run almost all games with a CPU filter on the 3DS. Master System games in particular look incredible (we no longer have to run them 'letterboxed' at 1:1 pixel scale for an optimal image). This is literally game changing.

Is it okay if I go ahead and make a pull request with your code snippet?

ekeeke commented:
Sure, go ahead. Just make sure to test all aforementioned core settings (without CPU filter) to see if they still work fine with these new max_width/max_height values, just in case I forgot about some edge case or got calculations wrong.

jdgleaver commented:
OK, I've cycled through all the core settings you mentioned with each of the following (CPU filter disabled):

- EUR/USA Game Gear games

- PAL/NTSC Master System games

- PAL/NTSC Genesis games (including Sonic 2, with Interlace Mode 2)

Everything seems fine, so I'll make the PR.

ekeeke commented:
Did you also tested following combinations that should usemaximal screen size ?

 - any Game Gear game with config.region forced to PAL, config.overscan = 3, config.gg_extra = 1 and config.ntsc = 1
 - any Master System game with config.region forced to PAL, config.overscan = 3 and config.ntsc = 1
 - Sonic 2 VS mode with config.region forced to PAL, config.overscan set to 3 and config.ntsc set enabled

jdgleaver commented:
Yes, I cycled through every combination (it was tedious...)

ekeeke commented:
OK, that's great then.

twinaphex commented:
Alright, sorry about the indiscretion. Next time around I will await a review by @ekeeke before merging. Glad to see that a middle ground could be reached though and that there don't seem to be major problems.
